### PR TITLE
Dynamic Configuration Refactoring

### PR DESCRIPTION
**Pull request by ldez**
**!!! Breaking changes !!!**

### What does this PR do

This PR introduces changes on the dynamic configuration.
- frontend becomes router
- backend becomes service
- new middleware approach (order customization, reusable configuration, chains)
- rules are now striclty matching the request without tweaking it (no more modifiers)

Tooling
- introduces a context based logger

### Motivation

A more flexible configuration file that gives additional possibilities to the end user
Logs that provide context to the end user

### Additional Notes

The only provider that has been migrated is `file`, as a proof of concept. **All other providers are broken**.

This Pull Request is the result of the work of many, and more specifically (in alphabetical order) @geraldcroes, @jbdoumenjou, @juliens, and @ldez

### CODE REVIEW

**Comment by emilevauge** [comment]
2lxjy5

File: log/log.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+	logger, ok := ctx.Value(loggerKey).(Logger)
+	if !ok {
+		logger = mainLogger
+	}
+
+	return logger
+}
+
+// WithoutContext Get main logger
```
Comment (Ullaakut) [inline log/log.go]: `gets`*

File: middlewares/accesslog/field_middleware.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	name    string
+	value   string
+	applyFn FieldApply
+}
+
+// NewFieldHandler creates a Field handler.
+func NewFieldHandler(next http.Handler, name string, value string, applyFn FieldApply) http.Handler {
+	return &FieldHandler{next: next, name: name, value: value, applyFn: applyFn}
+}
+
```
Comment (Ullaakut) [inline middlewares/accesslog/field_middleware.go]: Needs comment since exported

File: middlewares/accesslog/logger_formatters.go [hunk 3]
```diff
@@ -14,10 +14,10 @@ const (
 	defaultValue        = "-"
 )
 
-// CommonLogFormatter provides formatting in the Traefik common log format
+// CommonLogFormatter provides formatting in the Træfik common log format
```
Comment (Ullaakut) [inline middlewares/accesslog/logger_formatters.go]: Should be changed to `Traefik`

File: middlewares/accesslog/logger_formatters.go [hunk 4]
```diff
@@ -14,10 +14,10 @@ const (
...
 type CommonLogFormatter struct{}
 
-// Format formats the log entry in the Traefik common log format
+// Format formats the log entry in the Træfik common log format
```
Comment (Ullaakut) [inline middlewares/accesslog/logger_formatters.go]: Should be changed to `Traefik`

File: middlewares/accesslog/parser_test.go [hunk 5]
```diff
@@ -27,8 +27,8 @@ func TestParseAccessLog(t *testing.T) {
 				RequestRefererHeader:   `"testReferer"`,
 				RequestUserAgentHeader: `"testUserAgent"`,
 				RequestCount:           "1",
-				FrontendName:           `"testFrontend"`,
-				BackendURL:             `"http://127.0.0.1/testBackend"`,
+				RouterName:             `"testFrontend"`,
```
Comment (Ullaakut) [inline middlewares/accesslog/parser_test.go]: `testFrontend` should probably be changed to `testRouter` for consistency

File: middlewares/accesslog/parser_test.go [hunk 6]
```diff
@@ -47,8 +47,8 @@ func TestParseAccessLog(t *testing.T) {
 				RequestRefererHeader:   `"-"`,
 				RequestUserAgentHeader: `"Go-http-client/1.1"`,
 				RequestCount:           "1",
-				FrontendName:           `"testFrontend with space"`,
-				BackendURL:             `-`,
+				RouterName:             `"testFrontend with space"`,
```
Comment (Ullaakut) [inline middlewares/accesslog/parser_test.go]: `testFrontend` should probably be changed to `testRouter` for consistency

File: middlewares/accesslog/parser_test.go [hunk 7]
```diff
@@ -27,8 +27,8 @@ func TestParseAccessLog(t *testing.T) {
 				RequestRefererHeader:   `"testReferer"`,
 				RequestUserAgentHeader: `"testUserAgent"`,
 				RequestCount:           "1",
-				FrontendName:           `"testFrontend"`,
-				BackendURL:             `"http://127.0.0.1/testBackend"`,
+				RouterName:             `"testFrontend"`,
+				ServiceURL:             `"http://127.0.0.1/testBackend"`,
```
Comment (Ullaakut) [inline middlewares/accesslog/parser_test.go]: `testBackend` should probably be changed to `testService` for consistency

File: middlewares/addprefix/add_prefix.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// AddPrefix is a middleware used to add prefix to an URL request
+type addPrefix struct {
+	next   http.Handler
+	prefix string
+	name   string
+}
+
+// New creates a new New handler
```
Comment (Ullaakut) [inline middlewares/addprefix/add_prefix.go]: new new new new

File: middlewares/addprefix/add_prefix.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+			name:   name,
+		}
+	} else {
+		return nil, fmt.Errorf("prefix cannot be empty")
+	}
+
+	return result, nil
+}
+
+func (ap *addPrefix) GetTracingInformations() (string, ext.SpanKindEnum) {
```
Comment (Ullaakut) [inline middlewares/addprefix/add_prefix.go]: Should have comment since exported

File: middlewares/addprefix/add_prefix.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	return ap.name, tracing.SpanKindNoneEnum
+}
+
+func (ap *addPrefix) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (Ullaakut) [inline middlewares/addprefix/add_prefix.go]: Should have comment since exported

File: middlewares/addprefix/add_prefix_test.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/containous/traefik/config"
+	"github.com/containous/traefik/testhelpers"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewAddPrefix(t *testing.T) {
+	tests := []struct {
```
Comment (Ullaakut) [inline middlewares/addprefix/add_prefix_test.go]: Should be renamed to `testCases` for constistency

File: middlewares/addprefix/add_prefix_test.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestAddPrefix(t *testing.T) {
+	logrus.SetLevel(logrus.DebugLevel)
+	tests := []struct {
```
Comment (Ullaakut) [inline middlewares/addprefix/add_prefix_test.go]: Should be renamed to `testCases` for consistency

File: middlewares/auth/auth.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+	return append(users, appendUsers...), nil
+}
+
+func getLinesFromFile(filename string) ([]string, error) {
+	dat, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	// Trim lines and filter out blanks
```
Comment (Ullaakut) [inline middlewares/auth/auth.go]: newline missing after `if` statement

File: log/log_test.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	"bytes"
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_Log(t *testing.T) {
```
Comment (Ullaakut) [inline log/log_test.go]: Should be renamed to `TestLog` for consistency

File: middlewares/auth/basic_auth_test.go [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+	body, err := ioutil.ReadAll(res.Body)
+	require.NoError(t, err)
+	err = res.Body.Close()
+	require.NoError(t, err)
+
+	assert.Equal(t, "traefik\n", string(body))
+}
+
+func TestBasicAuthUsersFromFile(t *testing.T) {
+	tests := []struct {
```
Comment (Ullaakut) [inline middlewares/auth/basic_auth_test.go]: Should be renamed to `testCases` for consistency

File: middlewares/auth/digest_auth_test.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	req.SetBasicAuth("test", "test")
+
+	res, err := client.Do(req)
+	require.NoError(t, err)
+
+	assert.Equal(t, http.StatusUnauthorized, res.StatusCode)
+}
+
+func TestDigestAuthUsersFromFile(t *testing.T) {
+	tests := []struct {
```
Comment (Ullaakut) [inline middlewares/auth/digest_auth_test.go]: Should be renamed to `testCases` for consistency

**Comment by bboreham** [comment]
It looks like the metric `BackendServerUpGauge` stopped being set in this PR - is that right?

**Pull request merged**